subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/executer/ClassPathMerger.java

```diff
@@ -0,0 +1,122 @@
+/*
+ * Copyright 2018 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.integtests.fixtures.executer;
+
+import com.google.common.io.ByteStreams;
+import org.apache.commons.io.IOUtils;
+import org.gradle.api.Transformer;
+import org.gradle.internal.UncheckedException;
+import org.gradle.internal.os.OperatingSystem;
+import org.gradle.util.CollectionUtils;
+
+import java.io.ByteArrayInputStream;
+import java.io.File;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.net.URI;
+import java.nio.file.Files;
+import java.util.Collections;
+import java.util.List;
+import java.util.zip.ZipEntry;
+import java.util.zip.ZipOutputStream;
+
+/**
+ * On Windows, sometimes we hit `The filename or extension is too long` error when starting a new process. This is because of Windows command's 32KB limitation.
+ * In this situation, we merged all classpath entries into one empty `classpath.jar` with `Class-Path:` entry containing all original entries.
+ *
+ * See <a href="https://docs.oracle.com/javase/tutorial/deployment/jar/downman.html">Adding Classes to the JAR File's Classpath</>
+ * And <a href="https://github.com/bazelbuild/bazel/commit/d9a7d3a789be559bd6972208af21adae871d7a44">A similar implementation in Bazel</>
+ */
+interface ClassPathMerger {
+    ClassPathMerger INSTANCE = OperatingSystem.current().isWindows() ? new WindowsClassPathMerger() : new DoNothingClassPathMerger();
+    // Actually 32KB, let's leave some margin
+    int WINDOWS_CLASSPATH_LENGTH_LIMITATION = 30000;
+
+    List<File> mergeClassPathIfNecessary(List<File> classPath);
+
+    class WindowsClassPathMerger implements ClassPathMerger {
+        @Override
+        public List<File> mergeClassPathIfNecessary(List<File> classPath) {
+            if (CollectionUtils.join(File.pathSeparator, classPath).length() < WINDOWS_CLASSPATH_LENGTH_LIMITATION) {
+                return classPath;
+            }
+            return mergedClassPath(classPath);
+        }
+
+        private List<File> mergedClassPath(List<File> classPath) {
+            try {
+                String maneifestContent = generateManifestContent(classPath);
+                File jar = jar(maneifestContent);
+                return Collections.singletonList(jar);
```

Comment:
It feels weird the output of this method is always a single File but express it returns a `List<File>`. Maybe the `Collections.singletonList` should be done in `mergeClassPathIfNecessary` instead.

Reply:
Done.